src/cryptographic_primitives/polynomial.rs

```diff
@@ -0,0 +1,371 @@
+use std::convert::TryFrom;
+use std::{fmt, iter, ops};
+
+use derivative::Derivative;
+
+use crate::elliptic::curves::traits::{ECPoint, ECScalar};
+use crate::BigInt;
+
+/// Polynomial of some degree `n`
+///
+/// Polynomial has a form: `f(x) = a_0 + a_1 * x^1 + ... + a_(n-1) * x^(n-1) + a_n * x^n`.
+///
+/// Coefficients `a_i` and indeterminate `x` are scalars in curve prime field,
+/// ie. their type is `ECScalar` implementor.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "P::Scalar: Clone"))]
+#[derivative(Debug(bound = "P::Scalar: fmt::Debug"))]
+pub struct Polynomial<P: ECPoint> {
+    coefficients: Vec<P::Scalar>,
+}
+
+impl<P> Polynomial<P>
+where
+    P: ECPoint,
+    P::Scalar: Clone,
+{
+    /// Constructs polynomial `f(x)` from list of coefficients `a`.
+    ///
+    /// ## Order
+    ///
+    /// `a[i]` should corresponds to coefficient `a_i` of polynomial `f(x) = ... + a_i * x^i + ...`
+    ///
+    /// ## Panics
+    ///
+    /// Panics if list of coefficients is empty
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// # use curv::cryptographic_primitives::Polynomial;
+    /// # use curv::elliptic::curves::traits::ECScalar;
+    /// use curv::elliptic::curves::p256::{GE, FE};
+    ///
+    /// let coefs = vec![ECScalar::new_random(), ECScalar::new_random()];
+    /// let poly = Polynomial::<GE>::from_coefficients(coefs.clone());
+    ///
+    /// assert_eq!(coefs, poly.coefficients());
+    /// ```
+    pub fn from_coefficients(coefficients: Vec<P::Scalar>) -> Self {
+        assert!(
+            !coefficients.is_empty(),
+            "coefficients must have at least one coefficient"
+        );
+        Self { coefficients }
+    }
+
+    /// Sample a random polynomial of given degree
+    ///
+    /// ## Example
+    /// ```rust
+    /// # use curv::cryptographic_primitives::Polynomial;
+    /// # use curv::elliptic::curves::traits::ECScalar;
+    /// use curv::elliptic::curves::p256::{GE, FE};
+    ///
+    /// let polynomial = Polynomial::<GE>::sample(3);
+    /// assert_eq!(polynomial.degree(), 3);
+    /// ```
+    ///
+    pub fn sample(degree: u16) -> Self {
```

Comment:
Perhaps we have to make sure we didn't sample zero as the most significant indeterminate? This can result in the polynomial having a wrong degree.

Reply:
Thanks for catch - I added `sample_exact` and `sample_exact_with_fixed_const_term` functions that produce polynomial of given degree, whereas `sample` and `sample_with_fixed_const_term` functions are allowed now to produce polynomial of degree less or equal to given degree